File: julia/2022/day_04.jl
function parse_ranges(line::AbstractString)::Tuple{UnitRange{UInt}, UnitRange{UInt}}
    m = match(r"(\d+)-(\d+),(\d+)-(\d+)", line)
    s1, e1, s2, e2 = (parse(UInt, c) for c in m.captures)
    s1:e1, s2:e2
end

function full_overlap(r1::UnitRange, r2::UnitRange)::Bool
    r1 ∩ r2 in [r1, r2]
end

function partial_overlap(r1::UnitRange, r2::UnitRange)::Bool
    length(r1 ∩ r2) > 0
end

function main()
    text = read("input/2022/input_04.txt", String)
    range_pairs = split(text) .|> parse_ranges
    for (p, has_overlap) in enumerate([full_overlap, partial_overlap])
        num_overlaps = count(has_overlap(a, b) for (a, b) in range_pairs)
        println("Part $p: $num_overlaps")
    end
end

main()

File: rust/2022/src/bin/day_04.rs
use aoc::Part;
use std::ops::RangeInclusive;

fn parse_ranges(line: &str) -> [RangeInclusive<u8>; 2] {
    let vals: Vec<_> = line
        .split(&['-', ','])
        .map(|s| s.parse::<u8>().unwrap())
        .collect();
    [vals[0]..=vals[1], vals[2]..=vals[3]]
}

fn full_overlap(a: &RangeInclusive<u8>, b: &RangeInclusive<u8>) -> bool {
    a.start() <= b.start() && b.end() <= a.end() || b.start() <= a.start() && a.end() <= b.end()
}

fn partial_overlap(a: &RangeInclusive<u8>, b: &RangeInclusive<u8>) -> bool {
    a.start() <= b.end() && b.start() <= a.end()
}

fn main() {
    let input = aoc::read_input("2022/input_04.txt");
    let range_pairs: Vec<_> = input.lines().map(parse_ranges).collect();
    for (part, check_fn) in [
        (Part::One, full_overlap as fn(_, _) -> bool),
        (Part::Two, partial_overlap as fn(_, _) -> bool),
    ] {
        println!(
            "{part}: {}",
            range_pairs.iter().filter(|[a, b]| check_fn(a, b)).count()
        );
    }
}
